### PR TITLE
Fix #3989: Fix several unsoundness problems related to variant refinement

### LINKED ISSUE #3989: Unsoundness when inheriting conflicting type parameters
This compiles without warnings in both Dotty master and Scala 2.12:

```scala
trait A[+X]
case class B[X](x: X) extends A[X]
class C[X](x: Any) extends B[Any](x) with A[X]
def f(a: A[Int]): Int = a match {
  case B(i) => i
  case _ => 0
}
f(new C[Int]("foo"))
```

There used to be unsoundness warnings but they are gone since (presumably) https://github.com/lampepfl/dotty/issues/3918 even though the pattern match is unsound.

Furthermore the following compiles without warnings:

```scala
trait A[+X] { def get: X }
case class B[X](x: X) extends A[X] { def get: X = x }
class C[X](x: Any) extends B[Any](x) with A[X]
def g(a: A[Int]): Int = a.get
g(new C[Int]("foo"))
```

In this case, which doesn't rely on pattern matching, the definition of `C` itself is unsound but Dotty accepts it anyway. Scala 2.12 rightfully complains:

```
scala> class C[X](x: Any) extends B[Any](x) with A[X]
<console>:14: error: overriding method get in trait A of type => X;
 method get in class B of type => Any has incompatible type
       class C[X](x: Any) extends B[Any](x) with A[X]
             ^
```

### PR DESCRIPTION
**Pull request by odersky**
This solves a number of unsoundness problems related to variant refinement

 - implementation not covering the full member type (#3989, 2nd part)
 - pattern matching unsoundness (#3989, 1st part)
 - a variant of above, involving type patterns instead of unapply
 - the unsound protected[this] escape hatch for variance checking

Some tests had to be updated or re-classified because they were unsound before. The collection strawman needs to be updated to conform to the new rules.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -594,12 +594,61 @@ object RefChecks {
...
+      def checkMemberTypesOK() = {
+
+        // First compute all member names we need to check in `membersToCheck`.
+        // We do not check
+        //  - types
+        //  - synthetic members or bridges
+        //  - members in other concrete classes, since these have been checked before
+        //    (this is done for efficiency)
+        //  - members in a prefix of inherited parents that all come from Java or Scala2
+        //    (this is done to avoid false negatives since Scala2's rules for checking are different)
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: "false negatives" = false errors from the typechecker? I'm used to "false positives" (which makes sense for error checking).

More importantly, this means that we should test that Scala 2 material indeed follows Scala 2 rules?

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 2]
```diff
@@ -594,12 +594,61 @@ object RefChecks {
...
+        val membersToCheck = mutable.Set[Name]()
+        val seenClasses = mutable.Set[Symbol]()
+        def addDecls(cls: Symbol): Unit =
+          if (!seenClasses.contains(cls)) {
+            seenClasses.+=(cls)
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: Why not `seenClasses += cls`? Has idiomatic style changed?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: No, hat was an editing breakage.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
That's what I feared - performance is dramatically worse for scala-library, slightly worse for everything else.

**Comment by odersky** [comment]
The collection strawman does not compile under this branch because it contains some patterns that are now recognized as unsound. Some have easy fixes. E.g. `ArrayOps` needs to exten `IndexedSeqOps`, not `SeqOps`. That was probably an oversight. The main cause of errors is `WithFilter` on maps and sets. Here we lose precision wrt `collection.WithFilter` which leads to type inconsistencies. /cc @julienrf.

I have disabled the tests involving strawman until this is fixed. We should re-enable them afterwards.

**Comment by odersky** [comment]
test performance please

**Comment by julienrf** [comment]
@odersky Do we have to change our implementation? I think the changes you’ve mentioned for `ArrayOps` are arleady done [here](https://github.com/scala/collection-strawman/pull/436/files#diff-4e06810090b9ccdcab0f812bec7edeafL30). I have to have a look at withFilter…

**Comment by odersky** [comment]
@julienrf Yes, the ArrayOps changes seem to be already done. But `WithFilter` needs to be fixed. Can you take this branch and see what needs to be done? It already contains the changes to hk-equality which was the previous blocker for compiling the strawman with dotty.

**Comment by odersky** [comment]
@julienrf Also, the later changes here could invalidate some pattern matches. This is because it propagates fewer types from scrutinees to pattern. To fix this, one needs to sometimes write an @unchecked type argument in the pattern. I have noted one such case in `IterableOps` which now fails.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
Now all variants of problems I could derive from #3989 seem to be fixed.

**Comment by Blaisorblade** [comment]
@odersky Just in case, is this example with `protected[this]` also fixed? https://github.com/scala/bug/issues/7093#issuecomment-367033481

**Comment by odersky** [comment]
@Blaisorblade The `protected[this]` problem is fixed with latest commit. @julienrf: One more thing to update for the collection strawman. Hint: Compiling with -language:Scala2 -migration will give more specific error messages.

**Comment by odersky** [comment]
Regarding performance, I am inclined to not do anything right now. compileStdLib is an outlier, and I see no obviously safe way to reduce the number of checks.

**Comment by odersky** [comment]
@olafurpg Here's a problem scala-fix could help with: Insert @uncheckedVariance at the right places if a definition is no longer variance correct according to the new rules. `dotc` tried to do it, but it's more complex than it looks, so it's currently disabled. See f551646, in particular the two TODO comments there.

**Comment by szeiger** [comment]
Effect of this on collection strawman (with failures both in the original design and more in the changes I merged yesterday in https://github.com/scala/collection-strawman/pull/468:

```
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/BitSet.scala:40:23
[error] 40 |  protected[this] type BitSetC = C
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type C occurs in invariant position in type  = C of type BitSetC
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/Iterable.scala:29:43
[error] 29 |  protected[this] def fromSpecificIterable(coll: Iterable[A]): IterableCC[A] = iterableFactory.from(coll)
[error]    |                                           ^^^^^^^^^^^^^^^^^
[error]    |covariant type A occurs in contravariant position in type strawman.collection.Iterable[A] of value coll
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/Iterable.scala:30:22
[error] 30 |  protected[this] def newSpecificBuilder(): Builder[A, IterableCC[A]] = iterableFactory.newBuilder[A]()
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type A occurs in contravariant position in type (): strawman.collection.mutable.Builder[A, strawman.collection.Iterable[A]] of method newSpecificBuilder
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/Iterable.scala:750:24
[error] 750 |    protected[this] def filtered = View.Filter(toIterable, p, isFlipped = false)
[error]     |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]     |covariant type A occurs in invariant position in type => strawman.collection.View.Filter[A] of method filtered
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/Iterable.scala:71:23
[error] 71 |  protected[this] type IterableCC[X] = CC[X]
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type CC occurs in invariant position in type  = CC of type IterableCC
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/Iterable.scala:90:43
[error] 90 |  protected[this] def fromSpecificIterable(coll: Iterable[A]): C
[error]    |                                           ^^^^^^^^^^^^^^^^^
[error]    |covariant type A occurs in contravariant position in type strawman.collection.Iterable[A] of value coll
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/Iterable.scala:110:22
[error] 110 |  protected[this] def newSpecificBuilder(): Builder[A, C]
[error]     |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]     |covariant type A occurs in contravariant position in type (): strawman.collection.mutable.Builder[A, C] of method newSpecificBuilder
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/Map.scala:18:52
[error] 18 |  override protected[this] def fromSpecificIterable(coll: Iterable[(K, V)]): MapCC[K, V] = mapFactory.from(coll)
[error]    |                                                    ^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type V occurs in contravariant position in type strawman.collection.Iterable[(K, V)] of value coll
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/Map.scala:19:31
[error] 19 |  override protected[this] def newSpecificBuilder(): mutable.Builder[(K, V), MapCC[K, V]] = mapFactory.newBuilder[K, V]()
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type V occurs in contravariant position in type (): strawman.collection.mutable.Builder[(K, V), strawman.collection.Map[K, V]] of method newSpecificBuilder
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/Map.scala:62:23
[error] 62 |  protected[this] type MapCC[K, V] = CC[K, V]
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type CC occurs in invariant position in type  = CC of type MapCC
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/SortedMap.scala:17:52
[error] 17 |  override protected[this] def fromSpecificIterable(coll: Iterable[(K, V)]): SortedMapCC[K, V] = sortedMapFactory.from(coll)
[error]    |                                                    ^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type V occurs in contravariant position in type strawman.collection.Iterable[(K, V)] of value coll
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/SortedMap.scala:18:31
[error] 18 |  override protected[this] def newSpecificBuilder(): mutable.Builder[(K, V), SortedMapCC[K, V]] = sortedMapFactory.newBuilder[K, V]()
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |  covariant type V occurs in contravariant position in type ():
[error]    |    strawman.collection.mutable.Builder[(K, V),
[error]    |      strawman.collection.SortedMap[K, V]
[error]    |    ] of method newSpecificBuilder
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/SortedMap.scala:29:23
[error] 29 |  protected[this] type SortedMapCC[K, V] = CC[K, V]
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type CC occurs in invariant position in type  = CC of type SortedMapCC
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/SortedSet.scala:22:23
[error] 22 |  protected[this] type SortedIterableCC[X] = CC[X]
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type CC occurs in invariant position in type  = CC of type SortedIterableCC
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala:29:22
[error] 29 |  protected[this] def elemTag: ClassTag[A]
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type A occurs in invariant position in type => ClassTag[A] of method elemTag
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala:39:52
[error] 39 |  override protected[this] def fromSpecificIterable(coll: strawman.collection.Iterable[A]): ImmutableArray[A] = ImmutableArray.from[A](coll)(elemTag)
[error]    |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type A occurs in contravariant position in type strawman.collection.Iterable[A] of value coll
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/ImmutableArray.scala:41:31
[error] 41 |  override protected[this] def newSpecificBuilder(): Builder[A, ImmutableArray[A]] = ImmutableArray.newBuilder[A]()(elemTag)
[error]    |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |  covariant type A occurs in contravariant position in type ():
[error]    |    strawman.collection.mutable.Builder[A,
[error]    |      strawman.collection.immutable.ImmutableArray[A]
[error]    |    ] of method newSpecificBuilder
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/IntMap.scala:172:52
[error] 172 |  override protected[this] def fromSpecificIterable(coll: strawman.collection.Iterable[(Int, T)]): IntMap[T] =
[error]     |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]     |covariant type T occurs in contravariant position in type strawman.collection.Iterable[(scala.Int, T)] of value coll
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/IntMap.scala:180:31
[error] 180 |  override protected[this] def newSpecificBuilder(): Builder[(Int, T), IntMap[T]] =
[error]     |  ^
[error]     |  covariant type T occurs in contravariant position in type ():
[error]     |    strawman.collection.mutable.Builder[(scala.Int, T),
[error]     |      strawman.collection.immutable.IntMap[T]
[error]     |    ] of method newSpecificBuilder
[error] 181 |    new ImmutableBuilder[(Int, T), IntMap[T]](empty) {
[error] 182 |      def addOne(elem: (Int, T)): this.type = { elems = elems + elem; this }
[error] 183 |    }
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/LazyList.scala:231:40
[error] 231 |  protected[this] def cons[T](hd: => T, tl: => CC[T]): CC[T]
[error]     |                                        ^^^^^^^^^^^^
[error]     |covariant type CC occurs in contravariant position in type => CC[T] of value tl
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/LazyList.scala:361:43
[error] 361 |  protected[this] def newCons[T](hd: => T, tl: => CC[T]): CC[T]
[error]     |                                           ^^^^^^^^^^^^
[error]     |covariant type CC occurs in contravariant position in type => CC[T] of value tl
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/LongMap.scala:164:52
[error] 164 |  override protected[this] def fromSpecificIterable(coll: strawman.collection.Iterable[(Long, T)]): LongMap[T] = {
[error]     |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]     |covariant type T occurs in contravariant position in type strawman.collection.Iterable[(scala.Long, T)] of value coll
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/LongMap.scala:171:31
[error] 171 |  override protected[this] def newSpecificBuilder(): Builder[(Long, T), LongMap[T]] =
[error]     |  ^
[error]     |  covariant type T occurs in contravariant position in type ():
[error]     |    strawman.collection.mutable.Builder[(scala.Long, T),
[error]     |      strawman.collection.immutable.LongMap[T]
[error]     |    ] of method newSpecificBuilder
[error] 172 |    new ImmutableBuilder[(Long, T), LongMap[T]](empty) {
[error] 173 |      def addOne(elem: (Long, T)): this.type = { elems = elems + elem; this }
[error] 174 |    }
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/Map.scala:155:54
[error] 155 |    override protected[this] def fromSpecificIterable(coll: collection.Iterable[(K, V)]): WithDefault[K, V] =
[error]     |                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]     |covariant type V occurs in contravariant position in type strawman.collection.Iterable[(K, V)] of value coll
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/Map.scala:158:33
[error] 158 |    override protected[this] def newSpecificBuilder(): Builder[(K, V), WithDefault[K, V]] =
[error]     |    ^
[error]     |    covariant type V occurs in contravariant position in type ():
[error]     |      strawman.collection.mutable.Builder[(K, V),
[error]     |        strawman.collection.immutable.Map.WithDefault[K, V]
[error]     |      ] of method newSpecificBuilder
[error] 159 |      Map.newBuilder().mapResult((p: Map[K, V]) => new WithDefault[K, V](p, defaultValue))
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/SortedMap.scala:99:54
[error] 99 |    override protected[this] def fromSpecificIterable(coll: strawman.collection.Iterable[(K, V)]): WithDefault[K, V] =
[error]    |                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |covariant type V occurs in contravariant position in type strawman.collection.Iterable[(K, V)] of value coll
[error] -- Error: /Users/szeiger/code/scala/collection-strawman/collections/src/main/scala/strawman/collection/immutable/SortedMap.scala:102:33
[error] 102 |    override protected[this] def newSpecificBuilder(): Builder[(K, V), WithDefault[K, V]] =
[error]     |    ^
[error]     |    covariant type V occurs in contravariant position in type ():
[error]     |      strawman.collection.mutable.Builder[(K, V),
[error]     |        strawman.collection.immutable.SortedMap.WithDefault[K, V]
[error]     |      ] of method newSpecificBuilder
[error] 103 |      SortedMap.newBuilder().mapResult((p: SortedMap[K, V]) => new WithDefault[K, V](p, defaultValue))
[error] 27 errors found
```

**Comment by smarter** [comment]
Once the collection-strawman is fixed, we should check the performance impact there too.

**Comment by szeiger** [comment]
Shouldn't these be equivalent?

```scala
sealed abstract class Foo {
  type A
  protected[this] def elemTag: ClassTag[A]
}
```

```scala
sealed abstract class Foo[+A] {
  protected[this] def elemTag: ClassTag[A]
}
```

The first one still compiles but the second one doesn't.

**Comment by julienrf** [comment]
The version you gave with a type member is equivalent to an invariant type parameter (not a covariant one)

**Comment by olafurpg** [comment]
Thanks for the ping @odersky Noted https://github.com/scalacenter/scalafix/issues/628

**Comment by Blaisorblade** [comment]
@odersky I'd say breaking `collection-strawman` is a regression, so we can't merge this in a Dotty release till collection-strawman is fixed to match, IIUC Dotty's release policy.

Worst-case we could also merge the code behind some (temporary) flag, to avoid the changes staying out of tree too long and bitrotting, but not sure it's worth it.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 3]
```diff
@@ -594,12 +594,83 @@ object RefChecks {
...
+        def addDecls(cls: Symbol): Unit =
+          if (!seenClasses.contains(cls)) {
+            seenClasses.addEntry(cls)
+            for (mbr <- cls.info.decls)
+              if (mbr.isTerm && !mbr.is(Synthetic | Bridge) && mbr.memberCanMatchInheritedSymbols &&
+                  !membersToCheck.contains(mbr.name))
+                membersToCheck.addEntry(mbr.name)
+            cls.info.parents.map(_.classSymbol)
+              .filter(_.is(AbstractOrTrait))
+              .dropWhile(_.is(JavaDefined | Scala2x))
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: Why `dropWhile` and not `filter`?

**Comment by liufengyun** [comment]
The following code is unsound, causing runtime exception:

```Scala
object Test extends App {
  trait A[+X](val x: X)
  class B extends A(5) with A("hello")

  def f(a: A[Int]): Int = a match {
    case b: B => b.x   // error
    case _ => 0
  }

  f(new B)
}
```

**Comment by liufengyun** [comment]
A slight complex case that type checks but throws runtime exception:

```Scala
 object Test extends App {
  trait A[+X](val x: X)
  class B[+X](val y: X) extends A[X](y)

  class C extends B(5) with A[String]
  def f(a: A[Int]): String = a match {
    case c: C => c.x   // error
    case _ => "hello"
  }
  f(new C)
}
```

**Comment by Blaisorblade** [comment]
Should we try to fast track this so that `collection-strawman` can be fixed?

I could address current review comments and move other unsound examples to new issues.

**Comment by odersky** [comment]
Seems we dropped the ball on this one...

**Comment by odersky** [comment]
The problems reported by @liufengyun are now fixed as well. @liufengyun do you want to review the last two commits?

**Comment by odersky** [comment]
> LGTM, cannot find another counter-example.

Good news!

**Comment by adriaanm** [comment]
Does this also close the unsoundness illustrated by @TomasMikula in https://github.com/scala/bug/issues/10822#issuecomment-381109874, where GADT matching is unsoundly allowed because a covariant occurrence of a type param is hidden by a type ascription where it only occurs invariantly?

**Comment by smarter** [comment]
@adriaanm I think the unsoundness in scalac comes from allowing this to compile:
```scala
class Foo[A]
private[this] class Bar[+A] extends Foo[A]
```
Disabling the variance checks for `private[this]` is not sound here. In Dotty your example compiles because the variance checks for the extends clause are completely missing, but once we add them we won't special case them for `private[this]`: https://github.com/lampepfl/dotty/issues/2973

**Comment by adriaanm** [comment]
Gah, right! I should’ve caught that. Thanks for pointing it out.

**Pull request merged**